### PR TITLE
Join Section | Frontend

### LINKED ISSUE #280: Join Section User interaction [Design document]
## Task Flow:

- 1. User visits the main site https://www.realdevsquad.com 
        - Since the feature is still under development we have to pass a query parameter `join=true`
- 2. The user finds a section to join with a button
- 3. On clicking the button the user is redirected to the join page of the Realdevsquad website.
- 4. Here the user has to fill in their details (Handled using my site feature being built by Vinayak Trivedi)
- 5. On completion of the flow of my site user is redirected to a form where they fill in RDS Related details
        - a. Introduction
        - b.  How many hours can they commit to RDS
        - c.  Why do they want to join RDS
        - d. Others (More fields could be added based on further discussion)
- 6. On completion of this form, they will get a link, which needs to be sent for verification (currently shared manually to Ankush)
        - 6.1. Once the form details are verified they will get a fresh discord link to join the server (sent manually by Ankush)
- 7. On the final page, the user will only see the link to share and a message to wait for the link if the details get verified

### LINKED ISSUE #286: [RFC] Storage of join section data
## Problem Description
Once the user enters his/her details in the join section a personalized link is created for them to share with the verifier. This link contains their data. The **Issue** here is whether to store the data after verification of the user or as soon as the user submits the form.

## Current flow
- user signs in using the GitHub login.
- redirected to new signup flow by @vinayak-trivedi 
- after the signup flow user is sent to the join section to enter the details.
- a user-specific link is generated which has his/her data.
- link shared to the verifier to verify and send the discord invite link.

**Note:** The issue is to store the data before verification or after verification.

### PR DESCRIPTION
**Pull request by RitikJaiswal75**
**Will be functional after https://github.com/Real-Dev-Squad/website-backend/issues/714 is closed**
**Link for the back-end PR :** https://github.com/Real-Dev-Squad/website-backend/pull/715
## Problem Statement
Once the User Signs in to Peal Dev Squad, He / She has to wait for the discord invite of the real dev squad. Currently, this process is done by Ankush through manual acquisition and takes a lot of time asking the same question to everyone who wants to join Real Dev Squad.

## Solution Discussed
Create a Join section for Real dev Squad where the questions are already mentioned, and the user answers them. Once they answer all the questions, generate a personalized link for them that contains the data they entered.

## Current status:
- [x] Front-end part of the collection and sending data to the backend
- [x] Backend Task to upload and fetch data 

## References to the solution Discussions:
- https://github.com/Real-Dev-Squad/website-www/issues/280
- https://github.com/Real-Dev-Squad/website-www/issues/286

## Points to reviewers:
- the Percy generate snapshots test will fail as I have made UI changes. That cannot be fixed currently.
- In case any other test fails or you find any issues with file changes please comment here.

## Expected Flow
- User comes to [Real Dev Squad website](httpw://realdevsquad.com).
- They click on sign in with GitHub.
- complete the new/old signup flow (Whichever is active at the time.)
- User now has some data on Real Dev Squad
- Now we ask the user to answer the questions that were earlier asked manually with the help of this join section
  - User is sent to location `realdevsquad.com/join.html`
  - Page 1: Start the join process
  - Page 2: Enter personal (Location) Details
  - Page 3: Enter their introduction
  - page 4: why do they want to join RDS and how did they hear about us
  - page 5: Preview of the data Enter
  - page 6: final page with next steps and a link that contains personal info.
- Now the user waits for the data to be verified and after which they get the joining link for a discord server. 

### Issue
Closes #280 
Closes #286 

### What is the change?

Making join section for user introduction and verification

### Is Development Tested?

- [x] Yes
- [ ] No

https://user-images.githubusercontent.com/57758447/192142540-dafc63b0-773f-4d00-8744-0a5505661167.mp4

### CODE REVIEW

File: js/join.js [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+whyRds.addEventListener('input', () => {
+  window.localStorage.setItem('whyRds', whyRds.value);
+  toggleNextButton();
+  dataValidator(whyRds, 100);
+});
+
+heardAbout.addEventListener('input', () => {
+  window.localStorage.setItem('heardAbout', heardAbout.value);
+  toggleNextButton();
+});
```
Comment (yesyash) [inline js/join.js:260]: can we extract this into a function that takes an array of elements as an argument and adds the `input event listener` to each element? 
This is so we avoid repeating the same code multiple times.

Comment (yesyash) [reply js/join.js:260]: we are currently updating the local storage as soon as a change in the input is made, I think we can debounce the callback function of the event listener to avoid making very frequent changes to our local store.

**Comment by yesyash** [comment]
image

- instead of showing red input boxes by default I think we should add the `incorrect-data` class only when the user stops typing (or) moves to the next input box. 
- I think it's better to have `Your skills` input as a `multi-select dropdown` rather than a `text` input field.

**Comment by RitikJaiswal75** [comment]
> image
> 
> * instead of showing red input boxes by default I think we should add the `incorrect-data` class only when the user stops typing (or) moves to the next input box.
> * I think it's better to have `Your skills` input as a `multi-select dropdown` rather than a `text` input field.

Yes, Yash the content for skills and the way to input them is still a point for discussion. This feature will work better when we have a certain set of predefined skill sets.

**Comment by rupal-2505** [comment]
#289 


> > image
 Thw spelling of  `Feilds` is wrong it should be`Fields`.

**Comment by RitikJaiswal75** [comment]
> #289
> 
> > > image
> > > Thw spelling of  `Feilds` is wrong it should be`Fields`.

Thankyou @rupal-2505 I Fixed it

File: css/join.css [hunk 2]
```diff
@@ -0,0 +1,162 @@
+@import url('https://fonts.googleapis.com/css2?family=Roboto:wght@100;300;400;500;700;900&display=swap');
```
Comment (yesyash) [inline css/join.css:1]: Please move this to the HTML document.

**Review by yesyash — CHANGES_REQUESTED** [review]

File: css/join.css (hunk 2 above)
Comment (RitikJaiswal75) [reply css/join.css:1]: Can I know the reason for this

Comment (yesyash) [reply css/join.css:1]: Please refer to this:
https://web.dev/font-best-practices/

File: js/join.js [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+  }
+  if (whyRdsPageChecker()) {
+    previewBtn.classList.remove('button-disabled');
+  } else {
+    previewBtn.classList.add('button-disabled');
+  }
+}
+
+function autoFillTheFields() {
+  city.value = window.localStorage.getItem('city');
```
Comment (ivinayakg) [inline js/join.js:177]: can create an abstract functions like `addToLocalStorage` and `getFromLocalStorage`

File: js/join.js [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+  }
+});
+
+previous3.addEventListener('click', () => {
+  window.localStorage.setItem('flowState', flowState.introductionPage);
+  selectPage();
+});
+
+previewBtn.addEventListener('click', () => {
+  if (whyRdsPageChecker()) {
```
[the remark this answers is not in the document]
Comment (RitikJaiswal75) [reply js/join.js:272]: each button has different functionality and a different logic not related, so I am not sure how to go about this.

[the remark this answers is not in the document]
Comment (yesyash) [reply js/join.js:276]: you can make a function that takes an element and a callback function and apply that callback to the element.

**Review by ivinayakg — APPROVED** [review]

**Review by rupal-2505 — DISMISSED** [review]

File: css/join.css [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+}
+
+.text-red {
+  color: var(--text-red);
+}
+
+.button-disabled {
+  background-color: var(--input-background) !important;
+  border: none;
+  color: var(--theme-background) !important;
```
Comment (Maheima) [inline css/join.css:95]: What is the reason behind putting '!important' here?

**Review by Maheima — CHANGES_REQUESTED** [review]

File: css/join.css (hunk 5 above)
Comment (RitikJaiswal75) [reply css/join.css:95]: because I am overwriting the styling for filled button

**Review by Maheima — APPROVED** [review]
LGTM

**Review by rupal-2505 — APPROVED** [review]
LGTM!

**Comment by RitikJaiswal75** [comment]
Closed by mistake

File: css/join.css [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+}
+
+.container ol li {
+  line-height: 2rem;
+}
+
+.progress {
+  font-family: 'Roboto';
+  font-weight: 500;
+  font-size: 18px;
```
Comment (rohan09-raj) [inline css/join.css:26]: Can we please use short hand `font` property?

File: css/join.css [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+  line-height: 21px;
+  margin: 0 0 2rem;
+}
+
+.info,
+.input-label,
+.input-label-dark {
+  font-family: 'Roboto';
+  font-weight: 300;
+  font-size: 16px;
```
Comment (rohan09-raj) [inline css/join.css:36]: Can we please use short hand property?

File: css/join.css [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+}
+
+.input-label-dark {
+  font-weight: 500;
+}
+
+.user-input {
+  font-family: 'Roboto';
+  font-weight: 300;
+  font-size: 16px;
```
Comment (rohan09-raj) [inline css/join.css:51]: Here too can we please use short-hand?

File: css/join.css [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+  border: none;
+  color: var(--theme-background);
+  cursor: not-allowed;
+  opacity: 60%;
+  border-radius: 9px;
+  width: 11rem;
+  height: 3rem;
+  font-family: 'Roboto', sans-serif;
+  font-weight: 700;
+  font-size: 16px;
```
Comment (rohan09-raj) [inline css/join.css:87]: Can we please use shorthand?

File: css/join.css [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+.button-filled,
+.button-outline {
+  background-color: var(--theme-background);
+  border-radius: 9px;
+  width: 11rem;
+  height: 3rem;
+  color: var(--text-white);
+  font-family: 'Roboto', sans-serif;
+  font-weight: 700;
+  font-size: 16px;
```
Comment (rohan09-raj) [inline css/join.css:109]: Can we please use shorthand?

File: join.html [hunk 11]
```diff
@@ -82,9 +89,9 @@
     </button>
     <nav>
       <div class="hamburger">
-        <span class="bar" />
-        <span class="bar" />
-        <span class="bar" />
+        <span class="bar"></span>
+        <span class="bar"></span>
+        <span class="bar"></span>
```
Comment (rohan09-raj) [inline join.html:94]: Why this change is made?

File: join.html [hunk 12]
```diff
@@ -149,31 +156,242 @@
...
+          ></textarea>
+          <div class="twin-button-container">
+            <button class="button-outline" id="previous2">Previous</button>
+            <button class="button-disabled nextBtn" id="next2">Next</button>
+          </div>
+        </div>
+      </section>
+      <section class="page4" id="page4">
+        <h1 class="greeting">
+          Let’s Help the verifier Understand 🤔 you better.
```
Comment (rohan09-raj) [inline join.html:293]: Typo : Help -> help

File: join.html [hunk 13]
```diff
@@ -149,31 +156,242 @@
...
+          </select>
+          <br />
+          <div class="twin-button-container">
+            <button class="button-outline" id="previous3">Previous</button>
+            <button class="button-disabled nextBtn" id="next3">Preview</button>
+          </div>
+        </div>
+      </section>
+      <section class="page5" id="page5">
+        <h1 class="greeting">Here's a Preview 👀 of the data you entered</h1>
```
Comment (rohan09-raj) [inline join.html:335]: Preview -> preview

File: join.html [hunk 14]
```diff
@@ -149,31 +156,242 @@
...
+        <div class="container">
+          <h2 class="progress">Step 4 of 4: Preview</h2>
+          <p class="info">Please validate before Submitting</p>
```
Comment (rohan09-raj) [inline join.html:338]: Submitting -> submitting

File: join.html [hunk 15]
```diff
@@ -149,31 +156,242 @@
...
+          <p class="user-input input-regular" id="previewState"></p>
+          <p class="input-label-dark">Your Country</p>
+          <p class="user-input input-regular" id="previewCountry"></p>
+          <p class="input-label-dark">Your Introduction</p>
+          <p class="user-input input-big" id="previewIntro"></p>
+          <p class="input-label-dark">Your Skills</p>
+          <p class="user-input input-regular" id="previewSkills"></p>
+          <p class="input-label-dark">Your Institution</p>
+          <p class="user-input input-regular" id="previewInstitution"></p>
+          <p class="input-label-dark">What you do For Fun</p>
```
Comment (rohan09-raj) [inline join.html:355]: For Fun -> for fun

File: join.html [hunk 16]
```diff
@@ -149,31 +156,242 @@
...
+          <p class="user-input input-big" id="previewForFun"></p>
+          <p class="input-label-dark">A fun fact about you</p>
+          <p class="user-input input-big" id="previewFunFact"></p>
+          <p class="input-label-dark">Why do you want to join RDS?</p>
```
Comment (rohan09-raj) [inline join.html:359]: Can we please use `Real Dev Squad` instead of `RDS`?

File: js/join.js [hunk 17]
```diff
@@ -0,0 +1,290 @@
...
+}
+
+function showPage(currentFlowState) {
+  for (let i = 0; i < pages.length; i++) {
+    if (i == currentFlowState) {
+      pages[i].classList.remove('hide-page');
+    } else {
+      pages[i].classList.add('hide-page');
+    }
+  }
```
Comment (rohan09-raj) [inline js/join.js:123]: 1. We can use `forEach` here
2. If else conditions can be simplified with the toggle property of classList
https://developer.mozilla.org/en-US/docs/Web/API/Element/classList

File: js/join.js [hunk 18]
```diff
@@ -0,0 +1,290 @@
...
+  return (
+    introduction.value.trim().split(' ').length > 100 &&
+    skills.value.trim().split(' ').length > 6 &&
+    college.value.trim().split(' ').length > 5 &&
+    forFun.value.trim().split(' ').length > 100 &&
+    funFact.value.trim().split(' ').length > 100
+  );
+}
+
+function whyRdsPageChecker() {
```
Comment (rohan09-raj) [inline js/join.js:145]: Can we please name the functions better? 
Instead of Checker word, we can use Validator

File: js/join.js [hunk 19]
```diff
@@ -0,0 +1,290 @@
...
+heardAbout.addEventListener('input', () => {
+  window.localStorage.setItem('heardAbout', heardAbout.value);
+  toggleNextButton();
+});
+
+//Button Enablers
+
+startBtn.addEventListener('click', () => {
+  window.localStorage.setItem('flowState', flowState.personalDetailsPage);
+  let currentFlowState = localStorage.getItem('flowState');
```
Comment (rohan09-raj) [inline js/join.js:240]: Can we make this as constant?

File: js/join.js [hunk 20]
```diff
@@ -0,0 +1,290 @@
...
+      localStorage.setItem('flowState', (currentFlowState += 1));
+      previewFiller();
+    }
+    selectPage();
+  });
+});
+
+submit.addEventListener('click', async () => {
+  let data = JSON.stringify(localStorage);
+  let method = 'POST';
```
Comment (rohan09-raj) [inline js/join.js:273]: Can we please make these as `const`?

**Review by rohan09-raj — COMMENTED** [review]
Reference from the video
Can we please add a pointer to `copy` button and a hover effect, so that the user knows that it is clickable?

File: join.html (hunk 11 above)
Comment (RitikJaiswal75) [reply join.html:94]: self-closing span broke the join section. you can check it out at realdevsquad.com/join.html

File: js/join.js (hunk 17 above)
Comment (RitikJaiswal75) [reply js/join.js:123]: I tried with toggle but it won't work. toggling will not hide the previous page in this scenario as I am using pages[I] for that I will have to write pages[i-1] to toggle on that

File: js/join.js [hunk 21]
```diff
@@ -0,0 +1,290 @@
...
+    next2.classList.add('button-disabled');
+    next2.classList.remove('button-filled');
+  }
+  if (whyRdsPageChecker()) {
+    previewBtn.classList.remove('button-disabled');
+    previewBtn.classList.add('button-filled');
+  } else {
+    previewBtn.classList.add('button-disabled');
+    previewBtn.classList.remove('button-filled');
+  }
```
[the remark this answers is not in the document]
Comment (RitikJaiswal75) [reply js/join.js:181]: It won't work with the toggle as all have different functionality. Also these enhancements will be made in the next PR as per the discussion.

+ the site is going to be changed to ember so please take it  into account

**Comment by RitikJaiswal75** [comment]
I deleted the fork that was linked to this pull request by mistake to the work is now continued in
https://github.com/Real-Dev-Squad/website-www/pull/291

File: join.html (hunk 11 above)
Comment (rohan09-raj) [reply join.html:94]: I can see that self closing span tags are actually being nested in DOM 
image

Comment (rohan09-raj) [reply join.html:94]: Interesting to debug, why is that so
